### PR TITLE
SNOW-743920: Use axios for http communication

### PR DESCRIPTION
**Pull request by sfc-gh-dprzybysz**
### Description
- Rewrite HTTP request with axios instead of urllib
- Handle axios errors with responses as urllib does

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: test/integration/testLargeResultSet.js [hunk 1]
```diff
@@ -250,3 +251,75 @@ describe('Large result Set Tests', function ()
...
+
+  before(async () => {
+    connection = testUtil.createConnection();
+    await testUtil.connectAsync(connection);
+    // setting ROWS_PER_RESULTSET causes invalid, not encoded chunks from GCP
+    // await testUtil.executeCmdAsync(connection, 'alter session set ROWS_PER_RESULTSET = 1000000');
+    await testUtil.executeCmdAsync(connection, 'alter session set USE_CACHED_RESULT = false;');
+    await testUtil.executeCmdAsync(connection, createTable);
+    await testUtil.executeCmdAsync(connection, populateData);
+    configureLogger('TRACE');
```
Comment (sfc-gh-pmotacki) [inline test/integration/testLargeResultSet.js:272]: This level should be probably reduced before merging

File: test/integration/testConnectionWithOCSP.js [hunk 2]
```diff
@@ -58,8 +58,12 @@ describe('Connection with OCSP test', function ()
         {
           connection.connect(function (err)
           {
-            assert.strictEqual(err.code, Errors.codes.ERR_SF_RESPONSE_FAILURE);
-            callback();
+            try {
+              assert.strictEqual(err.code, Errors.codes.ERR_SF_RESPONSE_FAILURE);
+              callback()
+            }catch (e){
```
Comment (sfc-gh-pmotacki) [inline test/integration/testConnectionWithOCSP.js:64]: This line should be reformated to add spaces.

File: lib/http/node.js [hunk 3]
```diff
@@ -48,7 +48,7 @@ NodeHttpClient.prototype.getAgentAndProxyOptions = function (url, proxy, mock)
 {
   const isHttps = Url.parse(url).protocol === 'https:';
   let agentClass;
-  let agentOptions = { keepAlive: true };
+  let agentOptions = { keepAlive: true, keepAliveMsecs: 60000};
```
Comment (sfc-gh-pmotacki) [inline lib/http/node.js:51]: Question: Why I propose so long keep alive?

File: lib/http/base.js [hunk 4]
```diff
@@ -95,24 +96,19 @@ HttpClient.prototype.request = function (options)
...
-        }
-        else
-        {
-          response.body = response.data.toString();
-          await options.callback(null, normalizeResponse(response), data.toString());
-        }
+    request = axios.request({...requestOptions, transformResponse: res => res}).then(response => {
+      response.body = response.data;
+
+      if (options.callback) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-dprzybysz) [reply lib/http/base.js]: Yes, we should - changed

File: lib/http/node.js (hunk 3 above)
Comment (sfc-gh-dprzybysz) [reply lib/http/node.js]: it's removed from this PR

File: test/integration/testConnectionWithOCSP.js (hunk 2 above)
Comment (sfc-gh-dprzybysz) [reply test/integration/testConnectionWithOCSP.js]: I reformated it - thanks

File: .github/workflows/build-test.yml [hunk 5]
```diff
@@ -191,7 +191,7 @@ jobs:
         runs-on: ubuntu-latest
         strategy:
             matrix:
-                image: [ 'nodejs-centos6-default', 'nodejs-centos7-node12' ]
+                image: [ 'nodejs-centos7-node12' ]
```
Comment (sfc-gh-pmotacki) [inline .github/workflows/build-test.yml:194]: That change is not in scope but it is your choice to add this in the same PR.

File: ci/container/test_component.sh [hunk 6]
```diff
@@ -102,6 +102,27 @@ if [[ -z "$GITHUB_ACTIONS" ]]; then
...
+  echo "[DEBUG] Installing newer node 12"
+  export NVM_DIR=`pwd`/nvm
+  cp -r /usr/local/nvm $NVM_DIR
+  source $NVM_DIR/nvm.sh
+  nvm install 12
+
+  echo "[INFO] Build is using node versions"
+  npm version
+fi
+
```
Comment (sfc-gh-pmotacki) [inline ci/container/test_component.sh:125]: That change is not in scope but it is your choice to add this in the same PR.

File: lib/services/sf.js [hunk 7]
```diff
@@ -567,6 +567,13 @@ function StateAbstract(options)
         json: requestOptions.json,
         callback: async function (err, response, body)
         {
+          if (err && err.response) { // axios returns error for not 2xx responses - let's unwrap it
+            response = err.response;
+            err = null;
+            response.statusCode = response.status;
+            body = response.data;
+            response.body = response.data;
+          }
```
Comment (sfc-gh-pmotacki) [inline lib/services/sf.js:576]: What do you think about extract this unwraping to some commons?

Comment (sfc-gh-dprzybysz) [reply lib/services/sf.js:576]: it's changing incoming parameters so it's not so easy - but eventually yes, we should have common method to normalize errors

File: .github/workflows/build-test.yml (hunk 5 above)
Comment (sfc-gh-pmotacki) [reply .github/workflows/build-test.yml:194]: ok

File: ci/container/test_component.sh (hunk 6 above)
Comment (sfc-gh-dprzybysz) [reply ci/container/test_component.sh]: it is in separate PR https://github.com/snowflakedb/snowflake-connector-nodejs/pull/573 which should be merged first

File: lib/services/sf.js (hunk 7 above)
Comment (sfc-gh-dprzybysz) [reply lib/services/sf.js]: extracted to common place

**Pull request closed without merging**